libnmstate/netapplier.py

```diff
@@ -120,19 +121,22 @@ def generate_ifaces_metadata(ifaces_desired_state, ifaces_current_state):
     )
 
 
-def sanitize_ethernet_state(ifaces_desired_state):
+def sanitize_ethernet_state(ifaces_desired_state, ifaces_current_state):
     """
     If auto-negotiation, speed and duplex settings are not provided, they need
     to be set to None to not override them with the values from the current
     settings since the current settings are read from the device state and not
     from the actual configuration.  This makes it possible to distiguish
     whether a user specified these values in the later configuration step.
     """
-    for iface_state in six.viewvalues(ifaces_desired_state):
-        ethernet = iface_state.setdefault("ethernet", {})
-        ethernet.setdefault("auto-negotiation", None)
-        ethernet.setdefault("speed", None)
-        ethernet.setdefault("duplex", None)
+    for ifname, iface_state in six.viewitems(ifaces_desired_state):
+        if ifname in ifaces_current_state:
```

Comment:
AFAICS this could be simplified to something like

```
if 'ethernet' in ifaces_current_state.get(ifname, {})
    ethernet = ...
```

Reply:
I massaged it a bit.
I think we should check against the type and not against a subsequent property.